File: src/main/java/de/bixilon/minosoft/data/text/BaseComponent.java
/*
 * Minosoft
 * Copyright (C) 2020 Moritz Zwerger
 *
 * This program is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with this program.If not, see <https://www.gnu.org/licenses/>.
 *
 * This software is not affiliated with Mojang AB, the original developer of Minecraft.
 */

package de.bixilon.minosoft.data.text;

import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonPrimitive;
import de.bixilon.minosoft.modding.event.events.annotations.Unsafe;
import de.bixilon.minosoft.protocol.protocol.ProtocolDefinition;
import de.bixilon.minosoft.util.hash.BetterHashSet;
import javafx.collections.ObservableList;
import javafx.scene.Node;

import javax.annotation.Nullable;
import java.text.CharacterIterator;
import java.text.StringCharacterIterator;
import java.util.ArrayList;

public class BaseComponent extends ChatComponent {
    private static final String LEGACY_RESET_SUFFIX = String.valueOf(ProtocolDefinition.TEXT_COMPONENT_SPECIAL_PREFIX_CHAR) + PostChatFormattingCodes.RESET.getChar();
    private final ArrayList<ChatComponent> parts = new ArrayList<>();

    public BaseComponent() {
    }

    public BaseComponent(String text) {
        this(null, text);
    }

    public BaseComponent(@Nullable ChatComponent parent, String text) {
        // legacy String
        StringBuilder currentText = new StringBuilder();
        RGBColor color = null;
        BetterHashSet<ChatFormattingCode> formattingCodes = new BetterHashSet<>();
        StringCharacterIterator iterator = new StringCharacterIterator(text);
        while (iterator.current() != CharacterIterator.DONE) {
            char c = iterator.current();
            iterator.next();
            if (c != ProtocolDefinition.TEXT_COMPONENT_SPECIAL_PREFIX_CHAR) {
                currentText.append(c);
                continue;
            }
            // check next char
            char nextFormattingChar = iterator.current();
            iterator.next();
            RGBColor nextColor = ChatColors.getColorByFormattingChar(nextFormattingChar);
            if (nextColor != null) {
                // color change, add text part
                if (!currentText.isEmpty()) {
                    this.parts.add(new TextComponent(currentText.toString(), color, formattingCodes));
                    currentText = new StringBuilder();
                }
                color = nextColor;
                formattingCodes = new BetterHashSet<>();
                continue;
            }
            ChatFormattingCode nextFormattingCode = ChatFormattingCodes.getChatFormattingCodeByChar(nextFormattingChar);
            if (nextFormattingCode != null) {
                if (!currentText.isEmpty()) {
                    this.parts.add(new TextComponent(currentText.toString(), color, formattingCodes));
                    currentText = new StringBuilder();
                    color = null;
                    formattingCodes = new BetterHashSet<>();
                }
                formattingCodes.add(nextFormattingCode);
                if (nextFormattingCode == PostChatFormattingCodes.RESET) {
                    // special rule here
                    if (!currentText.isEmpty()) {
                        // color change, add text part
                        this.parts.add(new TextComponent(currentText.toString(), color, formattingCodes));
                        currentText = new StringBuilder();
                    }
                    color = null;
                    formattingCodes = new BetterHashSet<>();
                }
            }
        }
        if (!currentText.isEmpty()) {
            this.parts.add(new TextComponent(currentText.toString(), color, formattingCodes));
        }
    }

    public BaseComponent(JsonObject json) {
        this(null, json);
    }

    @SuppressWarnings("unchecked")
    public BaseComponent(@Nullable TextComponent parent, JsonElement data) {
        MultiChatComponent thisTextComponent = null;
        if (data instanceof JsonObject json) {
            if (json.has("text")) {
                String text = json.get("text").getAsString();
                if (text.contains(String.valueOf(ProtocolDefinition.TEXT_COMPONENT_SPECIAL_PREFIX_CHAR))) {
                    // legacy text component
                    this.parts.add(new BaseComponent(text));
                    return;
                }
                RGBColor color;
                if (parent != null && parent.getColor() != null) {
                    color = parent.getColor();
                } else {
                    color = null;
                }
                if (json.has("color")) {
                    String colorString = json.get("color").getAsString();
                    if (colorString.startsWith("#")) {
                        // RGB
                        color = new RGBColor(colorString);
                    } else {
                        color = ChatColors.getColorByName(colorString);
                    }
                }
                BetterHashSet<ChatFormattingCode> formattingCodes;
                if (parent != null && parent.getFormatting() != null) {
                    formattingCodes = (BetterHashSet<ChatFormattingCode>) parent.getFormatting().clone();
                } else {
                    formattingCodes = new BetterHashSet<>();
                }
                if (json.has("bold")) {
                    formattingCodes.addOrRemove(PreChatFormattingCodes.BOLD, json.get("bold").getAsBoolean());
                }
                if (json.has("italic")) {
                    formattingCodes.addOrRemove(PreChatFormattingCodes.ITALIC, json.get("italic").getAsBoolean());
                }
                if (json.has("underlined")) {
                    formattingCodes.addOrRemove(PreChatFormattingCodes.UNDERLINED, json.get("underlined").getAsBoolean());
                }
                if (json.has("strikethrough")) {
                    formattingCodes.addOrRemove(PreChatFormattingCodes.STRIKETHROUGH, json.get("strikethrough").getAsBoolean());
                }
                if (json.has("obfuscated")) {
                    formattingCodes.addOrRemove(PreChatFormattingCodes.OBFUSCATED, json.get("obfuscated").getAsBoolean());
                }

                thisTextComponent = new MultiChatComponent(text, color, formattingCodes);
                if (json.has("clickEvent")) {
                    thisTextComponent.setClickEvent(new ClickEvent(json.get("clickEvent").getAsJsonObject()));
                }
                if (json.has("hoverEvent")) {
                    thisTextComponent.setHoverEvent(new HoverEvent(json.get("hoverEvent").getAsJsonObject()));
                }
            }

            if (thisTextComponent != null) {
                this.parts.add(thisTextComponent);
            }

            final TextComponent parentParameter = thisTextComponent == null ? parent : thisTextComponent;
            if (json.has("extra")) {
                JsonArray extras = json.getAsJsonArray("extra");
                extras.forEach((extra -> this.parts.add(new BaseComponent(parentParameter, extra))));
            }

            if (json.has("translate")) {
                this.parts.add(new TranslatableComponent(parentParameter, json.get("translate").getAsString(), json.getAsJsonArray("with")));
            }
        } else if (data instanceof JsonPrimitive primitive) {
            this.parts.add(new BaseComponent(parent, primitive.getAsString()));
        }
    }

    @Override
    public String toString() {
        return PostChatFormattingCodes.RESET.getANSI() + getANSIColoredMessage();
    }

    @Override
    public String getANSIColoredMessage() {
        StringBuilder builder = new StringBuilder();
        this.parts.forEach((chatPart -> builder.append(chatPart.getANSIColoredMessage())));
        return builder.toString();
    }

    @Override
    public String getLegacyText() {
        StringBuilder builder = new StringBuilder();
        this.parts.forEach((chatPart -> builder.append(chatPart.getLegacyText())));
        String string = builder.toString();
        if (string.endsWith(LEGACY_RESET_SUFFIX)) {
            string = string.substring(0, string.length() - LEGACY_RESET_SUFFIX.length());
        }
        return string;
    }

    @Override
    public String getMessage() {
        StringBuilder builder = new StringBuilder();
        this.parts.forEach((chatPart -> builder.append(chatPart.getMessage())));
        return builder.toString();
    }

    @Override
    public ObservableList<Node> getJavaFXText(ObservableList<Node> nodes) {
        this.parts.forEach((chatPart) -> chatPart.getJavaFXText(nodes));
        return nodes;
    }

    @Unsafe
    public ArrayList<ChatComponent> getParts() {
        return this.parts;
    }

    public BaseComponent append(ChatComponent component) {
        this.parts.add(component);
        return this;
    }

    public BaseComponent append(String message) {
        this.parts.add(new BaseComponent(message));
        return this;
    }

    public boolean isEmpty() {
        return this.parts.isEmpty();
    }
}

File: src/main/java/de/bixilon/minosoft/data/text/BaseComponent.kt
/*
 * Minosoft
 * Copyright (C) 2020-2023 Moritz Zwerger
 *
 * This program is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with this program. If not, see <https://www.gnu.org/licenses/>.
 *
 * This software is not affiliated with Mojang AB, the original developer of Minecraft.
 */

package de.bixilon.minosoft.data.text

import de.bixilon.kutil.cast.CastUtil.nullCast
import de.bixilon.kutil.cast.CastUtil.unsafeCast
import de.bixilon.kutil.json.JsonObject
import de.bixilon.kutil.json.JsonUtil.toJsonList
import de.bixilon.kutil.json.JsonUtil.toJsonObject
import de.bixilon.kutil.primitive.BooleanUtil.toBoolean
import de.bixilon.minosoft.data.language.translate.Translator
import de.bixilon.minosoft.data.text.events.click.ClickEvents
import de.bixilon.minosoft.data.text.events.hover.HoverEvents
import de.bixilon.minosoft.data.text.formatting.FormattingCodes
import de.bixilon.minosoft.data.text.formatting.TextFormatting
import de.bixilon.minosoft.data.text.formatting.color.ChatColors.toColor
import de.bixilon.minosoft.data.text.formatting.color.RGBColor
import de.bixilon.minosoft.protocol.protocol.ProtocolDefinition
import de.bixilon.minosoft.util.KUtil.format
import de.bixilon.minosoft.util.KUtil.toResourceLocation
import de.bixilon.minosoft.util.nbt.tag.NBTUtil.get
import javafx.collections.ObservableList
import javafx.scene.Node

class BaseComponent : ChatComponent {
    val parts: MutableList<ChatComponent> = mutableListOf()

    constructor(parts: MutableList<ChatComponent>) {
        this.parts += parts
    }

    constructor(vararg parts: Any?) {
        for (part in parts) {
            this.parts += part.format()
        }
    }

    constructor(translator: Translator? = null, parent: TextComponent? = null, json: Map<String, Any>, restricted: Boolean = false) {
        val color = json["color"]?.nullCast<String>()?.toColor() ?: parent?.color

        val formatting = parent?.formatting?.copy() ?: TextFormatting()

        json["bold"]?.toBoolean()?.let { formatting[FormattingCodes.BOLD] = it }
        json["italic"]?.toBoolean()?.let { formatting[FormattingCodes.ITALIC] = it }
        json["underlined"]?.toBoolean()?.let { formatting[FormattingCodes.UNDERLINED] = it }
        json["strikethrough"]?.toBoolean()?.let { formatting[FormattingCodes.STRIKETHROUGH] = it }
        json["obfuscated"]?.toBoolean()?.let { formatting[FormattingCodes.OBFUSCATED] = it }

        val font = json["font"]?.toResourceLocation() ?: parent?.font
        val clickEvent = json["clickEvent", "click_event"]?.toJsonObject()?.let { click -> ClickEvents.build(click, restricted) } ?: parent?.clickEvent
        val hoverEvent = json["hoverEvent", "hover_event"]?.toJsonObject()?.let { hover -> HoverEvents.build(hover, restricted) } ?: parent?.hoverEvent

        val text = json["text"]?.nullCast<String>() ?: ""

        val component = TextComponent(
            message = text,
            color = color,
            formatting = formatting,
            font = font,
            clickEvent = clickEvent,
            hoverEvent = hoverEvent,
        )


        fun parseExtra() {
            json["extra"].toJsonList()?.let {
                for (data in it) {
                    this += ChatComponent.of(data, translator, component, restricted)
                }
            }
        }

        if (text.indexOf(ProtocolDefinition.TEXT_COMPONENT_FORMATTING_PREFIX) != -1) {
            this += ChatComponent.of(text, translator, component, restricted)
            parseExtra()
            return
        }

        if (text.isNotEmpty()) {
            this += component
        }

        parseExtra()

        json["translate"]?.toString()?.let {
            val with: Array<ChatComponent> = json.with(translator, component, restricted) ?: emptyArray()
            val fallback = json["fallback"]?.toString()
            this += translator?.forceTranslate(it.toResourceLocation(), component, restricted, fallback, data = with.unsafeCast()) ?: ChatComponent.of(json["with"], translator, component, restricted)
        }
    }

    override fun getJson(): Any {
        if (parts.isEmpty()) {
            return emptyList<Any>()
        }
        if (parts.size == 1) {
            return parts.first().getJson()
        }
        val list = mutableListOf<Any>()
        for (part in parts) {
            list += part.getJson()
        }
        return list
    }

    override fun cut(length: Int) {
        if (length <= 0) {
            throw IllegalArgumentException("Can not cut <= 0: $length")
        }
        var remaining = length
        for (part in parts) {
            val partLength = part.length
            if (remaining - partLength < 0) {
                part.cut(remaining)
                break
            }
            remaining -= partLength
        }
    }

    override fun copy(): ChatComponent {
        val parts: MutableList<ChatComponent> = mutableListOf()
        for (part in this.parts) {
            parts += part.copy()
        }

        return BaseComponent(parts)
    }

    override val ansi: String
        get() {
            val builder = StringBuilder()
            for (part in parts) {
                builder.append(part.ansi)
            }
            return builder.toString()
        }

    override val legacy: String
        get() {
            val stringBuilder = StringBuilder()
            for (part in parts) {
                stringBuilder.append(part.legacy)
            }
            // ToDo: Remove §r suffix
            return stringBuilder.toString()
        }

    override val message: String
        get() {
            val stringBuilder = StringBuilder()
            for (part in parts) {
                stringBuilder.append(part.message)
            }
            return stringBuilder.toString()
        }

    override fun getJavaFXText(nodes: ObservableList<Node>): ObservableList<Node> {
        for (part in parts) {
            part.getJavaFXText(nodes)
        }
        return nodes
    }

    override fun obfuscate(): BaseComponent {
        for (part in parts) part.obfuscate(); return this
    }

    override fun bold(): BaseComponent {
        for (part in parts) part.bold(); return this
    }

    override fun strikethrough(): BaseComponent {
        for (part in parts) part.strikethrough(); return this
    }

    override fun underline(): BaseComponent {
        for (part in parts) part.underline(); return this
    }

    override fun italic(): BaseComponent {
        for (part in parts) part.italic(); return this
    }

    override fun setFallbackColor(color: RGBColor): BaseComponent {
        for (part in parts) part.setFallbackColor(color); return this
    }

    override fun toString(): String {
        return legacy
    }

    operator fun plusAssign(component: ChatComponent) {
        if (component.length == 0) {
            return
        }
        parts += component
    }

    operator fun plusAssign(text: Any?) {
        this += text.format()
    }

    private fun <T> MutableSet<T>.addOrRemove(value: T, addOrRemove: Boolean?) {
        if (addOrRemove == null) {
            return
        }
        if (addOrRemove) {
            this.add(value)
        } else {
            this.remove(value)
        }
    }

    override fun hashCode(): Int {
        return parts.hashCode()
    }

    override fun equals(other: Any?): Boolean {
        if (other === this) {
            return true
        }
        if (other !is BaseComponent) {
            return false
        }
        return parts == other.parts
    }

    override val length: Int
        get() {
            var length = 0
            for (part in parts) {
                length += part.length
            }
            return length
        }

    override fun getTextAt(pointer: Int): TextComponent {
        var pointer = pointer
        for (part in parts) {
            val length = part.length
            if (pointer <= length) {
                return part.getTextAt(pointer)
            }
            pointer -= length
        }
        throw IllegalArgumentException("Pointer out of bounds!")
    }

    override fun trim(): ChatComponent? {
        return when {
            parts.isEmpty() -> null
            parts.size == 1 -> parts.first().trim()
            else -> {
                val parts: MutableList<ChatComponent> = mutableListOf()
                for (part in this.parts) {
                    parts += part.trim() ?: continue
                }
                if (parts.isEmpty()) return null
                if (parts.size == 1) parts.first()
                if (this.parts.size == parts.size) return this
                return BaseComponent(parts)
            }
        }
    }

    private fun JsonObject.with(translator: Translator?, parent: TextComponent, restricted: Boolean): Array<ChatComponent>? {
        val with = this["with"]?.toJsonList() ?: return null
        val parts: MutableList<ChatComponent> = mutableListOf()
        for (part in with) {
            if (part == null) continue
            parts += ChatComponent.of(raw = part, translator, parent, restricted = restricted)
        }

        return parts.toTypedArray()
    }
}
